src/redecomp/transfer/SparseMatrixTransfer.cpp

```diff
@@ -0,0 +1,306 @@
+  SLIC_ERROR_ROOT_IF(parent_test_space_.GetVDim() != 1 || parent_trial_space_.GetVDim() != 1 ||
+    redecomp_test_space_.GetVDim() != 1 || redecomp_trial_space_.GetVDim() != 1, 
+    "Only one value per element (zero-th order) is currently supported.");
+}
+
+std::unique_ptr<mfem::HypreParMatrix> SparseMatrixTransfer::TransferToParallel(
+  const mfem::SparseMatrix& src
+) const
+{
+  // check sizing on src
+  SLIC_ERROR_ROOT_IF(src.Height() != redecomp_test_space_.GetVSize(),
+    "The number of rows in src must match the size of the test FiniteElementSpace.");
+  SLIC_ERROR_ROOT_IF(src.Width() != redecomp_trial_space_.GetVSize(),
+    "The number of columns in src must match the size of the trial FiniteElementSpace.");
+  
+  // we want a csr matrix.  make sure src is finalized
+  SLIC_ERROR_ROOT_IF(!src.Finalized(), "src must be finalized first.");
+
+  auto my_rank = getMPIUtility().MyRank();
+  auto n_ranks = getMPIUtility().NRanks();
+
+  // package the entries to send to other ranks
+  MPIArray<double> recv_matrix_data(&getMPIUtility());
+  MPIArray<int> recv_parent_local_row(&getMPIUtility());
+  MPIArray<int> recv_parent_local_col(&getMPIUtility());
+  MPIArray<int> recv_parent_col_ranks(&getMPIUtility());
+  {
+    MPIArray<double> send_matrix_data(&getMPIUtility());
+    MPIArray<int> send_parent_local_row_offsets(&getMPIUtility());
+    MPIArray<int> send_parent_local_col_offsets(&getMPIUtility());
+    MPIArray<int> send_parent_col_ranks(&getMPIUtility());
+    // guess size and preallocate
+    auto approx_size = 2 * src.NumNonZeroElems() / n_ranks;
+    for (int r{0}; r < n_ranks; ++r)
+    {
+      send_matrix_data[r].reserve(approx_size);
+      send_parent_local_row_offsets[r].reserve(approx_size);
+      send_parent_local_col_offsets[r].reserve(approx_size);
+      send_parent_col_ranks[r].reserve(approx_size);
+    }
+
+    auto src_I = src.GetI();
+    auto src_J = src.GetJ();
+    auto src_data = src.GetData();
+
+    auto& test_elem_offsets = redecomp_test_mesh_.getRedecompToParentElemOffsets();
+    auto& test_ghost_elems = redecomp_test_mesh_.getRedecompToParentGhostElems();
+    auto& trial_elem_offsets = redecomp_trial_mesh_.getRedecompToParentElemOffsets();
+
+    // test space = row space
+    auto test_parent_rank = 0;
+    auto test_ghost_ct = 0;
+    for (int i{0}; i < src.Height(); ++i)
+    {
+      const auto test_elem_id = i; // assumes 1 point per element
+      while (test_elem_id == test_elem_offsets[test_parent_rank + 1])
+      {
+        ++test_parent_rank;
+        test_ghost_ct = 0;
+      }
+      // skip unowned elements
+      if (test_ghost_ct < test_ghost_elems[test_parent_rank].size() 
+        && test_elem_id == test_ghost_elems[test_parent_rank][test_ghost_ct])
+      {
+        ++test_ghost_ct;
+        continue;
+      }
+      // trial space = col space
+      auto trial_parent_rank = 0;
+      for (int Ij{src_I[i]}; Ij < src_I[i+1]; ++Ij)
```

Comment:
I think some words explaining this algorithm would be welcome.

Reply:
@ebchin - Per @jamiebramwell 's comment, this is a complex algorithm that relies on an inherent understanding or familiarity with CSR arrays, some underlying knowledge of the structure of a hypre par matrix, and the mpi communication. Per her comment, please review for any extra explanation that may help elaborate on this algorithm.